starrSeq/alt.go

```diff
@@ -0,0 +1,121 @@
+			cellTypeMap[i.Bx] = i.Cluster
+		}
+	}
+
+	bedEntries := bed.Read(s.AltMapping)
+	for _, i := range bedEntries {
+		tree = append(tree, i)
+		allConstructs = append(allConstructs, i.Name)
+	}
+	selectTree = interval.BuildTree(tree)
+
+	files := strings.Split(s.InFile, ",")
+	if len(files) > 1 {
+		multipleGems = true
+	}
+	for _, bam := range files {
+		inChan, _ := sam.GoReadToChan(bam)
+		for i := range inChan {
+			//filter sam file for some basic quality metrics
+			if !sam.FilterByQuality(i, 1) {
+				continue
+			}
+			if !sam.FilterByCigar(i, "starrSeqIntrons") {
+				continue
+			}
+			overlap = interval.Query(selectTree, i, "any")
+			if len(overlap) != 1 {
+				continue
+			}
+			cbc, cbcFound, _ := sam.QueryTag(i, "CB")
+			umi, umiFound, _ := sam.QueryTag(i, "UB")
+			if !umiFound {
+				continue
+			}
+			if multipleGems && cbcFound {
+				cbcUpdate = fmt.Sprintf("%s_%d", cbc.(string), gemNumber)
+				cluster, found = cellTypeMap[cbcUpdate]
+				umiBxSlice = append(umiBxSlice, fmt.Sprintf("%s_%s", umi.(string), cbcUpdate))
+			} else if cbcFound {
+				cbcUpdate = cbc.(string)
+				cluster, found = cellTypeMap[cbcUpdate]
+				umiBxSlice = append(umiBxSlice, fmt.Sprintf("%s_%s", umi.(string), cbcUpdate))
+			}
+			switch {
+			case !cbcFound:
+				starrSeqRead = Read{Bx: "NA", UMI: umi.(string), Cluster: "undefined", Construct: overlap[0].(bed.Bed).Name}
+			case cbcFound && found:
+				starrSeqRead = Read{Bx: cbcUpdate, UMI: umi.(string), Cluster: cluster, Construct: overlap[0].(bed.Bed).Name}
+			case cbcFound && !found:
+				starrSeqRead = Read{Bx: cbcUpdate, UMI: umi.(string), Cluster: "undefined", Construct: overlap[0].(bed.Bed).Name}
+			}
+			starrSeqReadSlice = append(starrSeqReadSlice, starrSeqRead)
+		}
+	}
+	collapsedUmiSlice := collapseUmi(starrSeqReadSlice)
+	for _, i := range collapsedUmiSlice {
+		if i.Construct == "" {
+			fmt.Println("collapseUmiSlice empty")
+		}
+	}
+	r := ReadSliceAnalysisSettings{FuncSettings: s, ReadSlice: collapsedUmiSlice, AllConstructs: allConstructs, UmiBxSlice: umiBxSlice, AllCellTypes: allCellTypes,
+		CellTypeMap: cellTypeMap}
+	ReadSliceAnalysis(r)
+}
+
+// collapseUmi will take in a slice of Read and collapse Read entries by UMI, keeping one entry per UMI.
+func collapseUmi(ssrs []Read) []Read {
+	var read Read
+	var found bool
+	var bxUmi string
+	var outSlice []Read
+
+	umiMap := make(map[string]Read) //make a map that has cellBarcode/UMI as the key and the corresponding Read entry is the value
+	for _, i := range ssrs {        // loop through slice of Read
+		bxUmi = fmt.Sprintf("%s_%s", i.Bx, i.UMI) //append cell barcode and UMI together
+		read, found = umiMap[bxUmi]               //check to see if the bx/umi is already in the map
+		if !found {
+			umiMap[bxUmi] = i //if it's not, add it with the Read to the map
+		} else {
+			if read.Construct == i.Construct { // if its found, check to make sure the current bx/umi maps to the same construct as the one in the map
```

Comment:
I thought you had a function where "if multiple reads have the same bxUmi, then keep the longer read". Was that ever a function?

Reply:
From Seth: part of samFilter not cellrangerBam